packages/lib/src/vis/heatmap/HeatmapMesh.tsx

```diff
@@ -10,67 +10,33 @@ import { useAxisSystemContext } from '../..';
 import type { VisScaleType } from '../models';
 import VisMesh from '../shared/VisMesh';
 import { DEFAULT_DOMAIN } from '../utils';
-import type { ColorMap } from './models';
+import type { ColorMap, TextureSafeTypedArray } from './models';
 import { getDataTexture, getInterpolator, scaleDomain } from './utils';
 
 interface Props {
-  values: NdArray<Float32Array | Uint16Array>; // if `Uint16Array`, it must contain half floats
+  values: NdArray<TextureSafeTypedArray | Uint16Array>; // uint16 values are treated as half floats
   domain: Domain;
   scaleType: VisScaleType;
   colorMap: ColorMap;
   invertColorMap?: boolean;
   magFilter?: TextureFilter;
-  alphaValues?: NdArray<Float32Array | Uint16Array>; // if `Uint16Array`, it must contain half floats
+  alphaValues?: NdArray<TextureSafeTypedArray | Uint16Array>; // uint16 values are treated as half floats
   alphaDomain?: Domain;
 }
 
 const CMAP_SIZE = 256;
 
-const SCALE_SHADER: Record<ScaleType, string> = {
-  [ScaleType.Linear]: `
-    float scale(float value) {
-      return oneOverRange * (value - min);
-    }
-
-    bool isSupported(float value) {
-      return true;
-    }`,
-  [ScaleType.Log]: `
-    float scale(float value) {
-      return oneOverRange * (log(value) * oneOverLog10 - min);
-    }
-
-    bool isSupported(float value) {
-      return value > 0.;
-    }`,
-  [ScaleType.SymLog]: `
-    float symlog(float x) {
-      return sign(x) * log(1. + abs(x)) * oneOverLog10;
-    }
-
-    float scale(float value) {
-      return oneOverRange * (symlog(value) - min);
-    }
-
-    bool isSupported(float value) {
-      return true;
-    }`,
-  [ScaleType.Sqrt]: `
-    float scale(float value) {
-      return oneOverRange * (sqrt(value) - min);
-    }
-
-    bool isSupported(float value) {
-      return value >= 0.;
-    }`,
-  [ScaleType.Gamma]: `
-    float scale(float value) {
-      return pow(oneOverRange * (value - min), gammaExponent);
-    }
-
-    bool isSupported(float value) {
-      return true;
-    }`,
+const IMPL_SCALE: Record<ScaleType, string> = {
+  [ScaleType.Linear]: 'norm(value)',
+  [ScaleType.Log]: 'norm(log(value) * oneOverLog10)',
+  [ScaleType.SymLog]: 'norm(sign(value) * log(1. + abs(value)) * oneOverLog10)',
+  [ScaleType.Sqrt]: 'norm(sqrt(value))',
+  [ScaleType.Gamma]: 'pow(norm(value), gammaExponent)',
+};
```

Comment:
Not a huge fan of the refactoring, to be honest.

Yes, it is way shorter but it relies on knowing how the template string is interpolated in the shader below instead of using abstractions as before.

I can live with it, though.

Reply:
As suggested by Thomas, `gammaExponent` now defaults to `1` and the gamma correction is always applied. This means we no longer need the `norm` function and can simplify this record even more with only the relevant scaling code.